### PR TITLE
Assigning multiple roles and abilities

### PR DESCRIPTION
**Pull request by Jeroen-G**
(#164 followup)

Did I miss anything?

```php

Bouncer::allow($user)->to(['edit','delete']);
Bouncer::disallow($user)->to(['edit','delete']);
Bouncer::forbid($user)->to(['edit', 'delete']);
Bouncer::unforbid($user)->to(['edit', 'delete']);
Bouncer::allow('admin')->to(['edit', 'delete']);
Bouncer::disallow('admin')->to(['edit', 'delete']);
Bouncer::assign('admin')->to([$user1, $user2]);
Bouncer::retract('admin')->from([$user1, $user2]);
$user->allow(['edit', 'delete']);
$user->disallow(['edit', 'delete']);
$user->assign(['supermod', 'moderator']);
$user->retract(['supermod', 'moderator']);

```

### CODE REVIEW

**Comment by JosephSilber** [comment]
I don't have the time to look at this at the moment. Will try to look through it next week.

Thanks for submitting this again :+1:

**Comment by Jeroen-G** [comment]
Take your time :)

**Comment by JosephSilber** [comment]
I think it would make sense to add model abilities to this.

```php
Bouncer::allow($user)->to([
    'edit' => Post::class,
    'delete' => $post,
]);
```

File: src/Database/Concerns/HasRoles.php [hunk 1]
```diff
@@ -30,25 +30,37 @@ public function roles()
...
-     * @param  \Silber\Bouncer\Database\Role|string  $role
+     * @param  \Silber\Bouncer\Database\Role|string|array  $role
      * @return $this
      */
-    public function assign($role)
+    public function assign($roles)
     {
-        (new AssignsRole($role))->to($this);
+        if (is_array($roles)) {
+            foreach ($roles as $role) {
```
Comment (JosephSilber) [inline src/Database/Concerns/HasRoles.php]: Isn't this kind of defeating the purpose of wanting to run it all in a single query?

File: tests/MultipleAbilitiesTest.php [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+     * When set to true, queries are counted and reported for tests in this file.
+     * @var boolean
+     */
+    public $report = false;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        if($this->report) {
```
Comment (JosephSilber) [inline tests/MultipleAbilitiesTest.php]: Please add spaces after all your `if` keywords.

File: src/Database/Concerns/HasRoles.php (hunk 1 above)
Comment (Jeroen-G) [inline src/Database/Concerns/HasRoles.php]: I see I forgot to commit something. It was wrapped in a DB transaction.

**Comment by JosephSilber** [comment]
How does that help, and why would you want that?

**Comment by Jeroen-G** [comment]
> I think it would make sense to add model abilities to this.
> 
> Bouncer::allow($user)->to([
>     'edit' => Post::class,
>     'delete' => $post,
> ]);
I wrote a test for this, and it is supported.

> How does that help, and why would you want that?
It was my understanding that wrapping it in a transaction would execute the queries at the end in one go. This appears to be a wrong thought.

**Comment by Jeroen-G** [comment]
> I think it would make sense to add model abilities to this.
> 
> Bouncer::allow($user)->to([
>     'edit' => Post::class,
>     'delete' => $post,
> ]);

Possible with the new commits 😄. It takes 2 queries less doing it combined like this than doing it separately. At this point and to my knowledge, assigning multiple roles is the only function not saving queries.

**Comment by JosephSilber** [comment]
Bouncer supports Laravel 5.1, which doesn't have the `Arr::isAssoc` method.

You can add it to Bouncer's `Helper` class: https://github.com/JosephSilber/bouncer/blob/master/src/Helper.php

**Comment by Jeroen-G** [comment]
I believe it does: https://github.com/illuminate/support/blob/5.1/Arr.php#L293

**Comment by JosephSilber** [comment]
@Jeroen-G it was introduced in `5.1.41`. Bouncer supports all the way back to `5.1.20`.

**Comment by Jeroen-G** [comment]
I already added it to the helper class :)

**Comment by JosephSilber** [comment]
We should probably also allow adding multiple abilities for a single model:

```php
Bouncer::allow($user)->to(['view', 'edit',], Post::class);
```

**Comment by Jeroen-G** [comment]
You mean [like this](https://github.com/JosephSilber/bouncer/pull/167/files#diff-93c98da4e4863e1f108ab091aedd70f7R49)?

**Comment by JosephSilber** [comment]
I have completely rewritten this from the ground up in https://github.com/JosephSilber/bouncer/commit/b1011e3462e31d7dd902689e67d710a4a93498fd

Still, I merged this to give you credit for your work, since I used this as a starting point.

Thank you :heart:

**Comment by Jeroen-G** [comment]
Thank you for implementing!

**Pull request merged**